### ISSUE TITLE
Molotov and Incendiary grenade projectile positioning bug

### ISSUE THREAD

**Issue by micvbang** [markus-wa/demoinfocs-golang#32]
We use the entity id of grenade projectiles when tracking them. This works well for non-fire grenades. It seems that new entities are created for fire grenades they explode, meaning that we can't use the entity id for tracking them and keeping `GameState.grenadeProjectiles` up to date.

Currently, this means that fire grenades will not get removed from `GameState.grenadeProjectiles` once they explode, meaning that they will accumulate through the demo.

**Comment by markus-wa** [comment]
I think this is the cause for the strange fire nade trajectories in https://github.com/markus-wa/demoinfocs-golang/tree/master/examples/nade-trajectories (red ones).

**Comment by markus-wa** [comment]
Could you confirm whether this is fixed with cf76d54 on dev?

https://github.com/markus-wa/demoinfocs-golang/blob/dev/examples/nade-trajectories/nade_trajectories.jpg now looks good to me.

**Comment by markus-wa** [comment]
I essentially moved the deletion of projectiles from the game-events to `handlePacketEntities` (where entities are deleted).

**Comment by micvbang** [comment]
Yep, it seems to work here as well. Good job!

I think it's a pretty nice solution.  I was looking for a place where we decidedly know that an entity is deleted from the game place a few times. 

I do see your point that there now is an extra lookup in the `grenadeProjectiles` map for every entity that is deleted, regardless of whether it's a grenade projectile or not. I think it's worth considering how we could refactor it, but mostly because there now is code in multiple places in order to keep `grenadeProjectiles` updated.

Perhaps introducing something like an `onEntityDestroy`-event on `Parser`, calling it from `handlePacketEntities` on every entity destruction event and then moving the updating of internal parser state there?

**Comment by markus-wa** [comment]
Yeah I'm not worried about performance, just cleanliness.

We could add another property like the hack I did here: https://github.com/markus-wa/demoinfocs-golang/blob/master/sendtables/sendtables.go#L49

But IMO that's not much better.

I'd prefer to avoid having internal event-handlers - it makes it hard to follow the call tree. The msg handlers make sense / are required because they get queued up.

Another option would be something like a `EntitiyDeletedEvent` on the `ServerClass` but that seems like an overkill (this would be different from the other events because there's no reflection involved).

**Comment by micvbang** [comment]
That moves it to where all the other grenade logic happens. I like it!

**Comment by micvbang** [comment]
We should get this comment removed as well: https://github.com/markus-wa/demoinfocs-golang/blob/dev/datatables.go#L310

**Comment by micvbang** [comment]
I would also consider moving the registration of the OnDestroy event to the top of the function, so that it's immediately clear that it's removed from `Parser.grenadeProjectiles` when it's destroyed :slightly_smiling_face:
